flowetl/etl/etl/production_task_callables.py

```diff
@@ -98,13 +95,23 @@ def move_file_and_record_ingestion_state__callable(
     to_path = mount_paths[to_dir]
 
     file_name = dag_run.conf["file_name"]
+    cdr_type = dag_run.conf["cdr_type"]
```

Comment:
I keep seeing these and wondering why they're hidden in the dag run conf, and not top level parameters?

Reply:
Cos thats where they live - they do not exist at compile time...

Reply:
you trigger a DAG with `conf=conf` and then that conf is in the `dag_run` for each task. When the DAG is constructed they do not exist....

Reply:
Ah right. So this is to enable having a single etl dag, rather than a dag per cdr type?

Reply:
So the cdr_type could probably be removed (could be a default arg to all callables) but the file name cannot...

Reply:
This is an airflow-specific limitation, right? I agree with @greenape that I'm a little confused every time I see this. I'm sure it makes sense when you think it through, but I'm still slightly hazy on the flow of information and which information is available at which time. So we should remember to document this well because I'm sure it'll catch people out otherwise.

Reply:
I feel like I'm being obtuse here, but _why_ can't the filename go? For this to work it is based on the date + Cdr type, right?

Reply:
Opened #801 for this.

Reply:
DAGs get triggered with a dictionary `conf` this is then accessible from all `TaskInstances` (running tasks) - some information is only known at runtime date of the data, type of the cdr for this file, where to store temp tables - all this goes in the `conf` dict and is used at *runtime*...

Reply:
Things needed at *runtime* cannot be passed to the DAG as args to the python callables...they must be extracted from the `conf` dict.

Reply:
Sure, but isn't the date stuff exactly what airflow's execution date concept about?

Reply:
You are asking me to completely redesign what I have spent time and energy on! I have no intention of doing that....

Reply:
Also it does not make sense because we do not receive data at regular occurrences. Data may arrive 2 days after the date of the data - it may come 5 days at a time on a single day...the airflow's execution date concept works if you control the other side and know things are present but we do not know that!

Reply:
I get the irregular data flow thing, but aren't you explicitly _providing_ the execution date when calling `trigger_dag`? Or am I misunderstanding how `trigger_dag` works? (Very possible, the docs are a bit terse on that).

Reply:
No the execution date will be the time it was executed which is `now()` - if it was set to the date of the data then you could never retry the ingestion.

Reply:
Then to fall back on airflow's retry machinery you get back into the issue of the irregular data flow....

Reply:
The filename could go...

Reply:
It can go from the model but it can't go from the config because I need to know it to move it - only other way is to keep the pattern that we matched hanging around and substitute the date back into it but I feel it's easier to just do that once and pass it to the conf.

Reply:
That's helpful to know, thanks.

My next question then, is why not just clear the DagRun if it already exists, as part of the sense dag?

The wider purpose here is that 'retry' is intended to be a single 'move the file back to the ingest folder' thing, right?

Reply:
Could do that but this is not part of this PR - haven't even implemented the trigger logic properly.

Reply:
Also the retry is not automatic - it needs a person to look at the file that failed and to manually move it back to the dump location.

Reply:
@josh-gree tells me that a future PR which adds a little more meat to this section will enable a better discussion, so I'm alright to move on.